budget-binder-multiplatform-app/src/commonMain/kotlin/de/hsfl/budgetBinder/common/DataResponse.kt

```diff
@@ -1,8 +1,52 @@
 package de.hsfl.budgetBinder.common
 
+import de.hsfl.budgetBinder.domain.usecase.NavigateToScreenUseCase
+import de.hsfl.budgetBinder.presentation.Screen
+import de.hsfl.budgetBinder.presentation.event.UiEvent
+import de.hsfl.budgetBinder.presentation.flow.RouterFlow
+import de.hsfl.budgetBinder.presentation.flow.UiEventSharedFlow
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+
 sealed class DataResponse<T>(val data: T? = null, val error: ErrorModel? = null) {
     class Success<T>(data: T) : DataResponse<T>(data)
     class Error<T>(error: ErrorModel?, data: T? = null) : DataResponse<T>(data, error)
     class Loading<T>(data: T? = null) : DataResponse<T>(data)
     class Unauthorized<T>(error: ErrorModel? = null, data: T? = null) : DataResponse<T>(data, error)
 }
+
+fun <T> DataResponse<T>.handleDataResponse(
+    scope: CoroutineScope = CoroutineScope(Dispatchers.Unconfined + SupervisorJob()),
+    routerFlow: RouterFlow = RouterFlow(NavigateToScreenUseCase(), scope),
+    onSuccess: (T) -> Unit,
+    onError: ((ErrorModel) -> Unit)? = null,
+    onLoading: (() -> Unit)? = null,
+    onUnauthorized: (() -> Unit)? = null
+) = scope.launch {
+    when (this@handleDataResponse) {
+        is DataResponse.Error -> {
+            onError?.let {
+                onError(this@handleDataResponse.error!!)
+            } ?: UiEventSharedFlow.mutableEventFlow.emit(UiEvent.ShowError(this@handleDataResponse.error!!.message))
+        }
+        is DataResponse.Loading -> {
+            onLoading?.let {
+                onLoading()
+            } ?: UiEventSharedFlow.mutableEventFlow.emit(UiEvent.ShowLoading)
+        }
+        is DataResponse.Success -> {
+            UiEventSharedFlow.mutableEventFlow.emit(UiEvent.HideSuccess)
+            onSuccess(this@handleDataResponse.data!!)
```

Comment:
und als letzten Kommentar, wäre es vielleicht sinnvoll irgendwie die Möglichkeit zu sagen, ob es ein HiddenSuccess ist oder nicht? Dann kann man die Funktion auch für ShowSuccess benutzen. Oder ist das relativ überflüssig, weil ShowSuccess HiddenSuccess "überschreibt" und er somit die Success-Nachricht anzeigt?

Reply:
Sollte überflüssig sein, weil wie du schon meintest ShowSuccess nach HideSuccess in den Flow geschmissen wird und das so dann überschrieben wird